baseplate.go

```diff
@@ -0,0 +1,140 @@
+package baseplate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/reddit/baseplate.go/batcherror"
+	"github.com/reddit/baseplate.go/edgecontext"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/secrets"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// Config is a general purpose config for assembling a Baseplate server
+type Config struct {
+	Addr    string
+	Timeout time.Duration
+
+	Log     log.Config
+	Metrics metricsbp.Config
+	Secrets secrets.Config
+	Sentry  log.SentryConfig
+	Tracing tracing.Config
+}
+
+// Baseplate is the general purpose object that you build a Server on.
+type Baseplate interface {
+	io.Closer
+
+	Config() Config
+	EdgeContextImpl() *edgecontext.Impl
+	Secrets() *secrets.Store
+}
+
+// Server is the primary interface for baseplate servers.
+type Server interface {
+	Baseplate() Baseplate
+	Serve() error
+}
```

Comment:
`Server` should have a `Stop` or `StopServe` or whatever it's called in the interface that's called from shutdown handler for graceful shutdown.